numba/byteflow.py

```diff
@@ -0,0 +1,1139 @@
+        # Each state is tracing a basic block
+        while runner.pending:
+            _logger.debug("pending: %s", runner.pending)
+            state = runner.pending.pop()
+            if state not in runner.finished:
+                _logger.debug("stack: %s", state._stack)
+                first_encounter[state.pc_initial] = state
+                # Loop over the state until it is terminated.
+                while True:
+                    runner.dispatch(state)
+                    # Terminated?
+                    if state.has_terminated():
+                        break
+                    # Not yet?
+                    else:
+                        state.advance_pc()
+                        # Must the new PC be a new block?
+                        if self._is_implicit_new_block(state):
+                            state.split_new_block()
+                            break
+                _logger.debug("end state. edges=%s", state.outgoing_edges)
+                runner.finished.add(state)
+                out_states = state.get_outgoing_states()
+                runner.pending.extend(out_states)
+
+        # Complete controlflow
+        self._build_cfg(runner.finished)
+        # Prune redundant PHI-nodes
+        self._prune_phis(runner)
+        # Post process
+        for state in sorted(runner.finished, key=lambda x: x.pc_initial):
+            self.block_infos[state.pc_initial] = si = adapt_state_infos(state)
+            _logger.debug("block_infos %s:\n%s", state, si)
+
+        assert self.block_infos
+
+    def _build_cfg(self, all_states):
+        graph = CFGraph()
+        for state in all_states:
+            b = state.pc_initial
+            graph.add_node(b)
+        for state in all_states:
+            for edge in state.outgoing_edges:
+                graph.add_edge(state.pc_initial, edge.pc, 0)
+        graph.set_entry_point(0)
+        graph.process()
+        self.cfgraph = graph
+
+    def _prune_phis(self, runner):
+        # Find phis that are unused in the local block
+        _logger.debug("Prune PHIs".center(60, '-'))
+
+        # Compute dataflow for used phis and propagate
+
+        # 1. Get used-phis for each block
+        # Map block to used_phis
+        def get_used_phis_per_state():
+            used_phis = defaultdict(set)
+            phi_set = set()
+            for state in runner.finished:
+                used = set(state._used_regs)
+                phis = set(state._phis)
+                used_phis[state] |= phis & used
+                phi_set |= phis
+            return used_phis, phi_set
+
+        # Find use-defs
+        def find_use_defs():
+            defmap = {}
+            phismap = defaultdict(set)
+            for state in runner.finished:
+                for phi, rhs in state._outgoing_phis.items():
+                    if rhs not in phi_set:
+                        # Is a definition
+                        defmap[phi] = state
+                    # else:
```

Comment:
this isn't an else?! the def and phi map both get updated if the phi is a defn ?

Reply:
can be safely removed